dbt/adapters/redshift/connections.py

```diff
@@ -47,24 +54,210 @@ class RedshiftCredentials(PostgresCredentials):
     ra3_node: Optional[bool] = False
+    connect_timeout: int = 10
+    role: Optional[str] = None
+    sslmode: Optional[str] = None
+    sslcert: Optional[str] = None
+    sslkey: Optional[str] = None
+    sslrootcert: Optional[str] = None
+    application_name: Optional[str] = "dbt"
+    retries: int = 0  # this is in-built into redshift python connector
+
+    _ALIASES = {"dbname": "database", "pass": "password"}
 
     @property
     def type(self):
         return "redshift"
 
     def _connection_keys(self):
-        keys = super()._connection_keys()
-        return keys + ("method", "cluster_id", "iam_profile", "iam_duration_seconds")
+        return "host", "port", "user", "database", "schema"
+
+    @property
+    def unique_field(self) -> str:
+        return self.host
+
+
+class RedshiftConnectMethodFactory:
+    credentials: RedshiftCredentials
+
+    def __init__(self, credentials):
+        self.credentials = credentials
+
+    def get_connect_method(self):
+        method = self.credentials.method
+        # Support missing 'method' for backwards compatibility
+        if method == RedshiftConnectionMethod.DATABASE or method is None:
+            logger.debug("Connecting to Redshift using 'database' credentials")
+            # this requirement is really annoying to encode into json schema,
+            # so validate it here
+            if self.credentials.password is None:
+                raise dbt.exceptions.FailedToConnectException(
+                    "'password' field is required for 'database' credentials"
+                )
+
+            def connect():
+                c = redshift_connector.connect(
+                    host=self.credentials.host,
+                    database=self.credentials.database,
+                    user=self.credentials.user,
+                    password=self.credentials.password,
+                    port=self.credentials.port if self.credentials.port else 5439,
+                    auto_create=self.credentials.autocreate,
+                    db_groups=self.credentials.db_groups,
+                )
+                if self.credentials.role:
+                    c.cursor().execute("set role {}".format(self.credentials.role))
+                return c
+
+            return connect
+
+        elif method == RedshiftConnectionMethod.IAM:
+            if not self.credentials.cluster_id:
+                raise dbt.exceptions.FailedToConnectException(
+                    "Failed to use IAM method, 'cluster_id' must be provided"
+                )
+
+            if self.credentials.iam_profile is None:
+                return self._get_iam_connect_method_from_env_vars()
+            else:
+                return self._get_iam_connect_method_with_tmp_cluster_credentials()
+        else:
+            raise dbt.exceptions.FailedToConnectException(
+                "Invalid 'method' in profile: '{}'".format(method)
+            )
+
+    def _get_iam_connect_method_from_env_vars(self):
```

Comment:
explicitly passing aws keys is not required for auth via env var if your use case works with [boto3's default credential chain](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials). also this would support use of credential/config file based credentials.

Reply:
Fixed this as well.